### PR TITLE
Geometry_Engine - Add an Offset method

### LINKED ISSUE #1066: Geometry_Engine: Add an offset method
As in a header.

### PR DESCRIPTION
**Pull request by LMarkowski**
### Issues addressed by this PR
Closes #1066 

Adding offset methods. `Offset(Arc)`, `Offset(Line)`, `Offset(Circle)` and `Offset(Polyline)` are mine. `Offset(Polycurve)` was made by @KonradStolarski.
Offset is hard to implement because of it's ambiguous nature. In more complicated cases there is no one proper way how it should work. Our methods work well on simple forms and polygons but get messy in more complicated shapes. In my opinion for now it is enough for most of engineering applications.
To avoid mistakes we added warnings about occurring unwanted intersections and changes in shapes of curves. 

### Test files
<s>[Here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FBHoM%5FEngine%2FGeometry%5FEngine%2FGeometry%5FEngine%2DIssue1066%2DOffset) is test file of offset method. In this case default random tests at the top of the file are less important. They are working for the small values of offset only. </s> [Here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FBHoM%5FEngine%2FGeometry%5FEngine%2FModify) is updated test. In the first section are generic test for simple curves. In the second there are some tests of complex curves. In the third there are generic tests of Polyline and PolyCurve which in this case are kinda useless. I invite to take a look at edge cases and make some tests of your own to see how this version of offset works and suggest any changes.

### Changelog
- `Modify.Offset()` Modify methods added in the `Geometry_Engine` for `Line`, `Arc`, `Circle`, `Polyline` and `PolyCurve` classes
- `Modify.Extend()` private Modify methods added in the `Geometry_Engine` for `Line`, `Arc`, `Circle`, `Ellipse`, `NurbsCurve`, `Polyline`, `PolyCurve` and `ICurve` classes
- `Modify.ExtendTangent()` private Modify methods added in the `Geometry_Engine` for `Line`, `Arc`, `Circle`, `Ellipse`, `NurbsCurve`, `Polyline`, `PolyCurve` and `ICurve` classes
- `Modify.Trim()` private Modify methods added in the `Geometry_Engine` for `Line`, `Arc`, `Circle`, `Ellipse`, `NurbsCurve`, `Polyline`, `PolyCurve` and `ICurve` classes
- `Modify.TrimExtend()` private Modify method added in the `Geometry_Engine` for `List<ICurve>` class

### Additional comments

Methods  `Modify.Extend()`,  `Modify.ExtendTangent()`,  `Modify.Trim()` and  `Modify.TrimExtend()` are all made by Konrad and are used only in `Modify.Offset()`. They haven't been properly tested as they are not a matter of this issue. It'd take a solid amount of time to step into Konrad's shoes and match them with the rest of the code. I think it'll be better to push Offset now with rest of methods marked as private and labelled with appropriate comment. Then raise another issue about finishing them properly.

#### TL;DR
Offset is much more complicated than it seems to be 😉

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
FYI I've updated PR title for change log requirements.

File: Geometry_Engine/Modify/Extend.cs [hunk 1]
```diff
@@ -21,22 +21,208 @@
...
 {
     public static partial class Modify
     {
         /***************************************************/
-        /**** Public Methods - Curves                   ****/
+        /**** Private Methods - Curves                  ****/
         /***************************************************/
 
-        public static Line Extend(this Line curve, double start = 0.0, double end = 0.0)
+        private static Line Extend(this Line curve, double start = 0.0, double end = 0.0)
```
Comment (IsakNaslundBh) [inline Geometry_Engine/Modify/Extend.cs]: any reason this method was made private?

**Comment by IsakNaslundBh** [comment]
Without that much context and as a general comment I would prefer not to have a whole list of private methods like this. Would for sure work to make them public, either by:

- Testing (ofc needed anyway, but test that they are fit for purpose)
- Renaming or similar if the methods are not general purpose.

Looking at some of the methods they look as they could be useful outside the context of offset, so can not really see the reason to keep them private, except for lack of multi-purpose testing?

File: Geometry_Engine/Modify/Extend.cs (hunk 1 above)
Comment (pawelbaran) [reply Geometry_Engine/Modify/Extend.cs]: > any reason this method was made private?

image

**Comment by pawelbaran** [comment]
@IsakNaslundBh the `Offset` method is a pain to dev and test - we would rather merge it as soon as it works not to make this PR too large/complex. After that a new issue will follow with the aim to make the trim and extend methods public.

File: Geometry_Engine/Modify/Extend.cs (hunk 1 above)
Comment (IsakNaslundBh) [reply Geometry_Engine/Modify/Extend.cs]: But this specific method was already in the code and has been there for over 2 years. See no reason why this should be made private now?

Comment (LMarkowski) [reply Geometry_Engine/Modify/Extend.cs]: @pawelbaran not this time actually, that one already was in the code and I mindlessly made it private while replacing every other `public` with `private`

Comment (LMarkowski) [reply Geometry_Engine/Modify/Extend.cs]: Yes, @IsakNaslundBh, you're right. I'll fix it

Comment (pawelbaran) [reply Geometry_Engine/Modify/Extend.cs]: Ah, right - my bad!

**Comment by LMarkowski** [comment]
> So far all good, but could you please:
> 
> * [x]  commit the method for `Line`
> * [ ]  move the test file to the Offset _subfolder_
> * [ ]  finish the test file for `Polyline`

@pawelbaran, I'd really manage to go back to my desk and do a quick commit with no need of everybody knowing that I forgot about it 😄 🙈

File: Geometry_Engine/Modify/Trim.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            }
+            List<ICurve> result = new List<ICurve>();
+            result = curve.Curves;
+            result[0] = result[0].ITrim(start, 0);
+            result[result.Count - 1] = result[result.Count - 1].ITrim(0, end);
+            return new PolyCurve
+            {
+                Curves = result
+            };
+
```
Comment (FraserGreenroyd) [inline Geometry_Engine/Modify/Trim.cs]: Delete blank line

File: Geometry_Engine/Modify/TrimExtend.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                curve=curve.ITrim(startPoint, curve.IEndPoint());
+                sP = true;
+            }
+            if(endPoint.IsOnCurve(curve))
+            {
+                curve=curve.ITrim(curve.IStartPoint(), endPoint);
+                eP = true;
+            }
+            if(sP&&eP)
```
Comment (FraserGreenroyd) [inline Geometry_Engine/Modify/TrimExtend.cs]: Put some spacing around `&&`

File: Geometry_Engine/Modify/TrimExtend.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                result.Add(curve);
+                return result;
+            }
+            if(sP&&!eP)
```
Comment (FraserGreenroyd) [inline Geometry_Engine/Modify/TrimExtend.cs]: Put some spacing around `&&`

File: Geometry_Engine/Modify/Offset.cs [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+            List<Point> Pts = new List<Point>();
+            List<Point[]> intersections = new List<Point[]>();
+
+            for (int i = 0; i < crvs.Count; i++)
+            {
+                intersections.Add(new Point[2]);
+                intersections[i][0] = crvs[i].IStartPoint();
+                intersections[i][1] = crvs[i].IEndPoint();
+            }
+            //intersecting arcs with arcs and lines
```
[the remark this answers is not in the document]
Comment (LMarkowski) [reply Geometry_Engine/Modify/Offset.cs]: Honestly, I don't know... It's Konrad's code and I did not check it line by line, as it works good overall.

[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply Geometry_Engine/Modify/Offset.cs]: This sounds like a good time to read Konrad's code and try to work it out then :wink: ideally the comments should bring something to the table to aid future developers, if neither of us can figure it out then maybe it should be removed - but I suspect it refers to either the code above it, or the code below it. If it's the code below it, then the blank line needs to be above it and the lower blank line should be deleted. If it's the other way round then the comment should probably go above the block of code it's for. If that makes sense.

**Comment by FraserGreenroyd** [comment]
I am happy with the code changes on this, however, there is still the following missing as far as I can see:
 - clear action plan for when the testing and subsequent improvement/making public of those methods will be conducted, including an issue raised detailing what needs to be done, and the milestone in which it will be achieved

This PR also now suffers from conflicts in the `csproj` file due to the time it's been open - this will need resolving. Many methods exist to do it, but given the complexity of this PR, I would recommend using the BHoM way of moving this to a new branch that is taken off `master` and reraising the PR from there once you've migrated the changes and checked they still work. @IsakNaslundBh would you agree?

**Comment by pawelbaran** [comment]
> I am happy with the code changes on this, however, there is still the following missing as far as I can see:
> 
> * clear action plan for when the testing and subsequent improvement/making public of those methods will be conducted, including an issue raised detailing what needs to be done, and the milestone in which it will be achieved
> 
This PR was a pain for a long time and further improvement will rather be a low priority - I would be fine with raising issues without a clear plan on when will they be resolved. Would that make you very unhappy @FraserGreenroyd?

**Comment by FraserGreenroyd** [comment]
> Would that make you very unhappy @FraserGreenroyd?

Not unhappy enough to block the PR :smile:

**Comment by LMarkowski** [comment]
Replacing with #1314
